nexus/db-queries/src/db/datastore/db_metadata.rs

```diff
@@ -8,21 +8,183 @@ use super::DataStore;
 use crate::db;
 use crate::db::error::public_error_from_diesel_pool;
 use crate::db::error::ErrorHandler;
-use async_bb8_diesel::AsyncRunQueryDsl;
+use async_bb8_diesel::{AsyncRunQueryDsl, AsyncSimpleConnection};
+use chrono::Utc;
 use diesel::prelude::*;
 use omicron_common::api::external::Error;
 use omicron_common::api::external::SemverVersion;
+use omicron_common::nexus_config::SchemaConfig;
+use slog::Logger;
+use std::collections::BTreeSet;
+use std::ops::Bound;
 use std::str::FromStr;
+use uuid::Uuid;
+
+/// Represents the ability of this Nexus to update the database schema.
+///
+/// Should not be constructed unless this Nexus instance successfully modified
+/// the db_metadata table to identify that a schema migration is in progress.
+pub struct SchemaUpdateLease {
+    our_id: Uuid,
+}
 
 impl DataStore {
+    // Ensures that the database schema matches "desired_version"
+    pub async fn ensure_schema(
```

Comment:
There is an assumption here that "all running nexus instances" at a new version are booting together. This is how we can read the schema version, and update to a "new version" without worrying about pulling the rug out from underneath a prior instance.

This isn't true! It may be possible that we run Nexus 1.0.0, and concurrently provision Nexus 2.0.0. This will be necessary follow-up work for online update, but this PR as-is relies on the commands from https://github.com/oxidecomputer/omicron/issues/3718 being executed.

To mitigate: I think Nexus is going to need to rely on tracking "what Nexus instances are provisioned, and where".

Reply:
"Booting together" means "are at the same version and/or are identical"--is that right?

Any reason not to put this note into a comment in the code?

Reply:
I will move this comment to the code -- just wanted to highlight these shortcomings and confirm they're "okay enough" for this to land in the short-term.

NOTE TO SELF BEFORE MARKING RESOLVED: File an issue for this

Reply:
I added a binary `schema-updater`, and removed the 'auto-format-on-boot' configuration from Nexus.

Reply:
Filed https://github.com/oxidecomputer/omicron/issues/3794